packages/uni-virtualizer/src/lib/layouts/Layout1dBase.ts

```diff
@@ -0,0 +1,428 @@
+    }
+  }
+
+  /**
+   * Amount of space in between items.
+   */
+  get spacing(): number {
+    return this._spacing;
+  }
+  set spacing(px) {
+    if (px !== this._spacing) {
+      this._spacing = px;
+      this._scheduleReflow();
+    }
+  }
+
+  /**
+   * Height and width of the viewport.
+   */
+  get viewportSize(): Size {
+    return this._viewportSize;
+  }
+  set viewportSize(dims) {
+    const {_viewDim1, _viewDim2} = this;
+    Object.assign(this._viewportSize, dims);
+    if (_viewDim2 !== this._viewDim2) {
+      this._viewDim2Changed();
+    } else if (_viewDim1 !== this._viewDim1) {
+      this._checkThresholds();
+    }
+  }
+
+  /**
+   * Scroll offset of the viewport.
+   */
+  get viewportScroll(): Positions {
+    return this._latestCoords;
+  }
+  set viewportScroll(coords) {
+    Object.assign(this._latestCoords, coords);
+    const oldPos = this._scrollPosition;
+    this._scrollPosition = this._latestCoords[this._positionDim];
+    if (oldPos !== this._scrollPosition) {
+      this._scrollPositionChanged(oldPos, this._scrollPosition);
+    }
+    this._checkThresholds();
+  }
+
+  /**
+   * Perform a reflow if one has been scheduled.
+   */
+  reflowIfNeeded() {
+    if (this._pendingReflow) {
+      this._pendingReflow = false;
+      this._reflow();
+    }
+  }
+
+  /**
+   * Scroll to the child at the given index, and the given position within that
+   * child.
+   */
+  scrollToIndex(index, position = 'start') {
+    if (!Number.isFinite(index))
+      return;
+    index = Math.min(this.totalItems, Math.max(0, index));
+    this._scrollToIndex = index;
+    if (position === 'nearest') {
+      position = index > this._first + this._num / 2 ? 'end' : 'start';
+    }
+    switch (position) {
+      case 'start':
+        this._scrollToAnchor = 0;
+        break;
+      case 'center':
+        this._scrollToAnchor = 0.5;
+        break;
+      case 'end':
+        this._scrollToAnchor = 1;
+        break;
+      default:
+        throw new TypeError(
+            'position must be one of: start, center, end, nearest');
+    }
+    this._scheduleReflow();
+    this.reflowIfNeeded();
+  }
+
+  async dispatchEvent(...args) {
+    await this._eventTargetPromise;
+    this._eventTarget.dispatchEvent(...args);
+  }
+
+  async addEventListener(...args) {
+    await this._eventTargetPromise;
+    this._eventTarget.addEventListener(...args);
+  }
+
+  async removeEventListener(...args) {
+    await this._eventTargetPromise;
+    this._eventTarget.removeEventListener(...args);
+  }
+
+  /**
+   * Get the top and left positioning of the item at idx.
+   */
+  abstract _getItemPosition(idx: number): Positions;
+
+  /**
+   * Update _first and _last based on items that should be in the current
+   * range.
+   */
+  abstract _getActiveItems();
+
+  updateItemSizes(sizes: {[key: number]: ItemBox}) {
+    // Override
+  }
+
+  protected _itemDim2Changed() {
+    // Override
+  }
+
+  protected _viewDim2Changed() {
+    // Override
```

Comment:
We should make it clearer that this method is lifecycle-callback-like. It is optional to implement.